constraint-solver/src/boolean_extractor.rs

```diff
@@ -28,6 +32,83 @@ pub fn extract_boolean<T: FieldElement, V: Ord + Clone + Hash + Eq>(
     Some(right + &(QuadraticSymbolicExpression::from_unknown_variable(z) * offset))
 }
 
+/// Tries to simplify a sequence of constraints by transforming them into affine
+/// constraints that make use of a new variable that is assumed to be boolean constrained.
+/// NOTE: The boolean constraint is not part of the output.
```

Comment:
I think an example, similar to the test, would help me here. Also, I always forget, how do we get rid of the boolean variables again? 😬 I think we still have them when reasoning about differences of addresses, but because their range is small, we know that the address does not overlap. But I'm a bit shaky on it and maybe here's a good place to explain it further.

Reply:
Yeah exactly, we currently do not get rid of the boolean variables. We just create an equivalent system we use for reasoning equivalence and non-equivalence. So the boolean variables do not end up in the system we are optimizing.

How do we get rid of them during reasoning: We do an exhaustive search.